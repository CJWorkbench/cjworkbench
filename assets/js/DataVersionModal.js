import React from 'react'
import PropTypes from 'prop-types'
import Modal from 'reactstrap/lib/Modal'
import ModalHeader from 'reactstrap/lib/ModalHeader'
import ModalBody from 'reactstrap/lib/ModalBody'
import ModalFooter from 'reactstrap/lib/ModalFooter'
import memoize from 'memoize-one'
import { setDataVersionAction, updateWfModuleAction } from './workflow-reducer'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'

const Months = [
  // AP style
  'Jan.',
  'Feb.',
  'March',
  'April',
  'May',
  'June',
  'July',
  'Aug.',
  'Sept.',
  'Oct.',
  'Nov.',
  'Dec.',
]

/**
 * Return AP-style date in the user's time zone, but with the date before the
 * time.
 *
 * For instance: "June 22, 2018 – 10:35 a.m."
 */
function formatDate(date) {
  const mon = Months[date.getMonth()]
  const dd = date.getDate()
  const yyyy = date.getFullYear()
  let hh = date.getHours()
  const mm = String(100 + date.getMinutes()).slice(1) // 0-padded
  let ampm = 'a.m.'

  if (hh === 0) {
    hh = 12
  } else if (hh === 12) {
    ampm = 'p.m.'
  } else if (hh > 12) {
    ampm = 'p.m.'
    hh -= 12
  }

  return `${mon} ${dd}, ${yyyy} – ${hh}:${mm} ${ampm}`
}

/**
 * Form <input type="radio">. Calls onSelect on change.
 */
class FetchVersion extends React.PureComponent {
  static propTypes = {
    id: PropTypes.string.isRequired, // version ID
    date: PropTypes.instanceOf(Date).isRequired, // version date
    isSelected: PropTypes.bool.isRequired,
    isSeen: PropTypes.bool.isRequired, // has user selected this version ever
    onSelect: PropTypes.func.isRequired, // func(versionId) => undefined
  }

  onChange = (ev) => {
    if (ev.target.checked) {
      this.props.onSelect(this.props.id)
    }
  }

  render() {
    const { id, date, isSeen, isSelected } = this.props

    let className = isSeen ? 'seen' : 'unseen'
    if (isSelected) className += ' selected'

    return (
      <label className={className}>
        <input type="radio" name="data-version" value={id} checked={isSelected} onChange={this.onChange} />
        <time time={date.toISOString()}>{formatDate(date)}</time>
      </label>
    )
  }
}

/**
 * Form that calls onSubmit(wantNotifications).
 *
 * Implemented as a checkbox with an onChange event. ([2018-06-22] it _looks_
 * like a <button>, but it's a checkbox in a <label>.)
 */
class NotificationsForm extends React.PureComponent {
  static propTypes = {
    notificationsEnabled: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired, // func(bool) => undefined
  }

  onChange = (ev) => {
    this.props.onSubmit(ev.target.checked) // should change the state
  }

  render() {
    const checked = this.props.notificationsEnabled

    const className = checked ? 'notifications-enabled' : 'notifications-disabled'

    return (
      <form onSubmit={this.onSubmit} className={`notifications ${className}`}>
        <p className="status"><i className="icon icon-notification"/> Alerts are <strong>{checked ? 'on' : 'off'}</strong></p>
        <p className="description">{ checked ? (
          'If the output of this module changes, I will receive an email'
        ) : (
          'Turn alerts ON to receive an email if the output of this module changes'
        )}</p>
        <label>
          <input name="notifications-enabled" type="checkbox" checked={checked} onChange={this.onChange}/>
          <span className="action">{checked ? 'Turn off' : 'Turn on'}</span>
        </label>
      </form>
    )
  }
}

export class DataVersionModal extends React.PureComponent {
  static propTypes = {
    fetchWfModuleId: PropTypes.number.isRequired,
    fetchWfModuleName: PropTypes.string.isRequired,
    fetchVersions: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      date: PropTypes.instanceOf(Date).isRequired,
      isSeen: PropTypes.bool.isRequired,
    })).isRequired,
    selectedFetchVersionId: PropTypes.string.isRequired,
    wfModuleId: PropTypes.number.isRequired,
    notificationsEnabled: PropTypes.bool.isRequired, // whether enabled on selectedWfModule
    onClose: PropTypes.func.isRequired, // func() => undefined
    onChangeFetchVersionId: PropTypes.func.isRequired, // func(wfModuleId, versionId) => undefined
    onChangeNotificationsEnabled: PropTypes.func.isRequired, // func(wfModuleId, isEnabled) => undefined
  }

  state = {
    selectedFetchVersionId: this.props.selectedFetchVersionId
  }

  onChangeNotificationsEnabled = (isEnabled) => {
    const { onChangeNotificationsEnabled, wfModuleId } = this.props
    onChangeNotificationsEnabled(wfModuleId, isEnabled)
  }

  setSelectedFetchVersionId = (selectedFetchVersionId) => {
    this.setState({ selectedFetchVersionId })
  }

  onSubmit = () => {
    if (this.state.selectedFetchVersionId !== this.props.selectedFetchVersionId) {
      const { fetchWfModuleId, onChangeFetchVersionId } = this.props
      onChangeFetchVersionId(fetchWfModuleId, this.state.selectedFetchVersionId)
    }

    this.props.onClose()
  }

  render() {
    const {
      fetchWfModuleName,
      fetchVersions,
      onClose,
      notificationsEnabled,
    } = this.props

    return (
      <Modal className="data-versions-modal" isOpen={true} fade={false} toggle={onClose}>
        <ModalHeader toggle={onClose}>Data Versions </ModalHeader>
        <ModalBody>
          <form onSubmit={this.onSubmit} onCancel={this.onClose}>
            <p className="introduction">These are the versions of “{fetchWfModuleName}” that we have stored:</p>
            <ol>
              {fetchVersions.map(v => <li key={v.id}><FetchVersion
                onSelect={this.setSelectedFetchVersionId}
                isSelected={this.state.selectedFetchVersionId === v.id}
                {...v}
                /></li>)}
            </ol>
          </form>
        </ModalBody>
        <ModalFooter>
          <NotificationsForm
            notificationsEnabled={notificationsEnabled}
            onSubmit={this.onChangeNotificationsEnabled}
            />
          <button
            name="load"
            disabled={this.state.selectedFetchVersionId === this.props.selectedFetchVersionId}
            onClick={this.onSubmit}
            >Load</button>
        </ModalFooter>
      </Modal>
    )
  }
}

const getWorkflow = ({ workflow }) => workflow
/**
 * Find first WfModule that has a `.loads_data` ModuleVersion.
 */
const getFetchWfModule = createSelector([ getWorkflow ], (workflow) => {
  return (workflow.wf_modules || []).find(wfModule => {
    return wfModule.module_version && wfModule.module_version.module && wfModule.module_version.module.loads_data
  }) || null
})

/**
 * Parse `wfModule.versions.versions` Array of { id, date, isSeen }.
 *
 * wfModule.versions.versions is an Array of [ dateString, isSeen ]
 * pairs.
 */
const getFetchVersions = memoize(versions => {
  return versions.map(version => {
    const [ id, isSeen ] = version
    return {
      id,
      isSeen,
      date: new Date(id),
    }
  })
})

function mapStateToProps(state, { wfModuleId }) {
  const fetchWfModule = getFetchWfModule(state)

  const wfModule = state.workflow.wf_modules.find(m => m.id === wfModuleId)
  const notificationsEnabled = wfModule ? wfModule.notifications : false

  return {
    fetchWfModuleId: fetchWfModule ? fetchWfModule.id : null,
    fetchWfModuleName: fetchWfModule ? fetchWfModule.module_version.module.name : null,
    fetchVersions: fetchWfModule ? getFetchVersions(fetchWfModule.versions.versions || []) : null,
    selectedFetchVersionId: fetchWfModule ? fetchWfModule.versions.selected : null,
    notificationsEnabled,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    onChangeFetchVersionId(wfModuleId, versionId) {
      dispatch(setDataVersionAction(wfModuleId, versionId))
    },

    onChangeNotificationsEnabled(wfModuleId, isEnabled) {
      dispatch(updateWfModuleAction(wfModuleId, { notifications: isEnabled }))
    },
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DataVersionModal)
