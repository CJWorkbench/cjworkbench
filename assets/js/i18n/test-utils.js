import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
