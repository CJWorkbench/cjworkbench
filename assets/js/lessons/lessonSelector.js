import { createSelector } from 'reselect'
import { StateWithHelpers } from '../lessons/DoneHelpers'
import { matchLessonHighlight } from '../util/LessonHighlight'

function isStepDone(sectionTitle, stepIndex, stateWithHelpers, step) {
  // Canonical example testJs:
  // `return workflow.selectedWfModule.moduleName === 'Add from URL'`
  const fn = new Function('state', 'workflow', '"use strict"; ' + step.testJs)
  // Give our function a name: makes it easy to debug crashes
  Object.defineProperty(fn, 'name', {
    value: `LessonSection "${sectionTitle}" Step ${stepIndex + 1}`,
    writable: false,
  })

  try {
    return fn(stateWithHelpers, stateWithHelpers.workflow)
  } catch (e) {
    console.error(e)
    console.error('The previous error is a bug in this function, called with these arguments:', fn, stateWithHelpers, stateWithHelpers.workflow)
    return false
  }
}

function calculateActiveStep(stateWithHelpers, sections) {
  // Run each testJs function until one returns false
  for (let sectionIndex = 0; sectionIndex < sections.length; sectionIndex++) {
    const section = sections[sectionIndex]
    const steps = section.steps || []
    for (let stepIndex = 0; stepIndex < steps.length; stepIndex++) {
      const step = steps[stepIndex]
      if (!isStepDone(section.title, stepIndex, stateWithHelpers, step)) {
        return {
          activeSectionIndex: sectionIndex,
          activeStepIndex: stepIndex,
          activeStep: step,
        }
      }
    }
  }

  // all steps complete
  return {
    activeSectionIndex: null,
    activeStepIndex: null,
    activeStep: null,
  }
}

/**
 * Return `false` always, so no component gets lesson-highlighted when there
 * is no lesson.
 */
function testLessonHighlightButThereIsNoLesson(test) { return false }

const getWorkflow = ({ workflow }) => workflow
const getTabs = ({ tabs }) => tabs
const getWfModules = ({ wfModules }) => wfModules
const getModules = ({ modules }) => modules
const getSelectedWfModule = ({ selected_wf_module }) => selected_wf_module
const getLessonData = ({ lessonData }) => lessonData || null

/**
 * Returns the Lesson for the given state.
 *
 * Example:
 *
 * ```
 * import lessonSelector from '../lessons/lessonSelector'
 * function mapStateToProps(state, ownProps) {
 *   const { testHighlight } = lessonSelector(state)
 *   return {
 *     isLessonHighlight: testHighlight({ type: 'WfModule', moduleName: ownProps.moduleName })
 *   }
 * }
 * ```
 *
 * The returned object has the following properties:
 *
 * * `activeSectionIndex`: section containing the step we have arrived at
 * * `activeStepIndex`: step within the section that the user needs to complete
 * * `testHighlight`: function that accepts a `LessonHighlightType` and
 *                          returns `true` iff the active step specifies we
 *                          should highlight it.
 *
 * This function never throws. Lessons may have invalid `testJs` values,
 * especially if they index into arrays and the user does something we don't
 * expect. Those errors will be caught and logged with console.error(). An
 * erroring step counts as unfinished: usually, errors mean we haven't imagined
 * a user doing something -- which probably means the user _didn't_ do what we
 * suggested.
 */
const getLesson = createSelector(
  [ getWorkflow, getTabs, getWfModules, getModules, getSelectedWfModule, getLessonData ],
  (workflow, tabs, wfModules, modules, selectedWfModule, lessonData) => {
    if (lessonData === null) {
      return {
        activeSectionIndex: null,
        activeStepIndex: null,
        testHighlight: testLessonHighlightButThereIsNoLesson
      }
    }

    const stateWithHelpers = new StateWithHelpers({
      workflow,
      tabs,
      wfModules,
      modules,
      selected_wf_module: selectedWfModule
    })
    const { activeSectionIndex, activeStepIndex, activeStep } = calculateActiveStep(stateWithHelpers, lessonData.sections)
    const lessonHighlight = activeStep ? activeStep.highlight : []
    const testHighlight = (test) => matchLessonHighlight(lessonHighlight, test)

    return {
      activeSectionIndex,
      activeStepIndex,
      testHighlight
    }
  })
export default getLesson
