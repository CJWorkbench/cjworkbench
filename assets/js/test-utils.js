/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
