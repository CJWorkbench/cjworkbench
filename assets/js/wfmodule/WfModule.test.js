/* globals describe, it, expect, jest, beforeEach, afterEach */
import React from 'react'
import ConnectedWfModule, { WfModule } from './WfModule'
import DataVersionModal from '../DataVersionModal'
import { okResponseMock } from '../test-utils'
import { shallow, mount } from 'enzyme'
import deepEqual from 'fast-deep-equal'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import lessonSelector from '../lessons/lessonSelector'

jest.mock('../lessons/lessonSelector', () => jest.fn()) // same mock in every test :( ... we'll live

describe('WfModule, not read-only mode', () => {
  let props
  let mockApi

  const buildShallowProps = () => ({
    isReadOnly: false,
    isAnonymous: false,
    isZenMode: false,
    name: 'TestModule',
    wfModule: wfModule,
    module: module,
    changeParam: jest.fn(),
    removeModule: jest.fn(),
    inputWfModule: { id: 123, last_relevant_delta_id: 707 },
    selected: true,
    api: mockApi,
    index: 2,
    onDragStart: jest.fn(),
    onDragEnd: jest.fn(),
    isLessonHighlight: false,
    isLessonHighlightCollapse: false,
    isLessonHighlightNotes: false,
    fetchModuleExists: false,
    clearNotifications: jest.fn(),
    setSelectedWfModule: jest.fn(),
    setWfModuleCollapsed: jest.fn(),
    setZenMode: jest.fn()
  })

  // A mock module that looks like LoadURL
  const wfModule = {
    'id': 999,
    'notes': '',
    'is_collapsed': false, // false because we render more, so better test
    'parameter_vals': [
      {
        'id': 100,
        'parameter_spec': {
          'id_name': 'url',
          'type': 'string'
        },
        'value': 'http://some.URL.me'
      },
      {
        'id': 101,
        'parameter_spec': {
          'id_name': 'version_select',
          'type': 'custom'
        }
      }
    ],
    module_version: { module: 2 }
  }

  const module = { id_name: 'loadurl', name: 'Load from URL' }

  beforeEach(() => {
    // Reset mock functions before each test
    mockApi = {
      postParamEvent: okResponseMock(),
      onParamChanged: okResponseMock(),
      setWfModuleNotes: okResponseMock()
    }

    props = buildShallowProps()
  })

  it('matches snapshot', () => {
    const wrapper = shallow(<WfModule {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('is has .status-busy', () => {
    const w = shallow(<WfModule {...props} wfModule={{...wfModule, status: 'busy'}} />)
    expect(w.hasClass('status-busy')).toBe(true)
    w.setProps({ wfModule: { ...wfModule, status: 'ready' } })
    w.update()
    expect(w.hasClass('status-busy')).toBe(false)
    expect(w.hasClass('status-ready')).toBe(true)
  })

  it('supplies getParamText and setParamText', () => {
    const wrapper = shallow(<WfModule {...props} />)
    const instance = wrapper.instance()

    expect(instance.getParamText('url')).toEqual('http://some.URL.me')
    wrapper.instance().setParamText('url', 'http://foocastle.ai')

    expect(props.changeParam).toHaveBeenCalledWith(100, { value: 'http://foocastle.ai' })
    // and the input prop should not be mutated
    expect(wfModule.parameter_vals[0].value).toEqual('http://some.URL.me')
  })

  it('renders a note', () => {
    const wrapper = shallow(<WfModule {...props} wfModule={Object.assign({}, wfModule, { notes: 'some notes' })} />)
    expect(wrapper.find('EditableNotes').prop('value')).toEqual('some notes')
  })

  it('renders in Zen mode', () => {
    const wrapper = shallow(<WfModule {...props} isZenMode />)
    expect(wrapper.prop('className')).toMatch(/\bzen-mode\b/)
    expect(wrapper.find('WfParameter').map(n => n.prop('isZenMode'))).toEqual([ true, true ])
  })

  it('has an "enter zen mode" button', () => {
    const aModule = { id_name: 'pythoncode', name: 'Python Code', has_zen_mode: true }
    const wrapper = shallow(<WfModule {...props} module={aModule} />)
    let checkbox = wrapper.find('input[type="checkbox"][name="zen-mode"]')
    expect(checkbox.prop('checked')).toBe(false)
    expect(wrapper.find('WfParameter').map(n => n.prop('isZenMode'))).toEqual([ false, false ])

    checkbox.simulate('change', { target: { checked: true } })
    expect(props.setZenMode).toHaveBeenCalledWith(wfModule.id, true)
    wrapper.setProps({ isZenMode: true })

    checkbox = wrapper.find('input[type="checkbox"][name="zen-mode"]')
    expect(checkbox.prop('checked')).toBe(true)
    expect(wrapper.find('WfParameter').map(n => n.prop('isZenMode'))).toEqual([ true, true ])

    checkbox.simulate('change', { target: { checked: false } })
    expect(props.setZenMode).toHaveBeenCalledWith(wfModule.id, false)
  })

  it('renders notifications, opening and closing a modal', () => {
    const wrapper = shallow(<WfModule {...props} fetchModuleExists isReadOnly={false} isAnonymous={false} />)

    wrapper.find('button.notifications').simulate('click')
    expect(props.clearNotifications).toHaveBeenCalledWith(999)
    expect(wrapper.find(DataVersionModal).length).toBe(1)

    wrapper.find(DataVersionModal).prop('onClose')()
    wrapper.update()
    expect(wrapper.find(DataVersionModal).length).toBe(0)
  })

  it('hides notifications when isAnonymous', () => {
    const wrapper = shallow(<WfModule {...props} fetchModuleExists isReadOnly={false} isAnonymous />)
    expect(wrapper.find('button.notifications').length).toBe(0)
  })

  it('hides notifications when isReadOnly', () => {
    const wrapper = shallow(<WfModule {...props} fetchModuleExists isReadOnly isAnonymous={false} />)
    expect(wrapper.find('button.notifications').length).toBe(0)
  })

  it('hides notifications when there is no fetch module', () => {
    const wrapper = shallow(<WfModule {...props} fetchModuleExists={false} isReadOnly={false} isAnonymous={false} />)
    expect(wrapper.find('button.notifications').length).toBe(0)
  })

  it('adds a note', () => {
    const wrapper = shallow(<WfModule {...props} />)

    expect(wrapper.find('.module-notes.visible')).toHaveLength(0)

    wrapper.find('button.edit-note').simulate('click')
    wrapper.find('EditableNotes').simulate('change', { target: { value: 'new note' } })
    wrapper.find('EditableNotes').simulate('blur')

    expect(wrapper.find('EditableNotes').prop('value')).toEqual('new note')
    expect(mockApi.setWfModuleNotes).toHaveBeenCalledWith(wfModule.id, 'new note')

    expect(wrapper.find('.module-notes.visible')).toHaveLength(1)
  })

  it('deletes a note', () => {
    const wrapper = shallow(<WfModule {...props} wfModule={Object.assign({}, wfModule, { notes: 'some notes' })} />)

    expect(wrapper.find('.module-notes.visible')).toHaveLength(1)

    wrapper.find('EditableNotes').simulate('change', { target: { value: '' } })
    wrapper.find('EditableNotes').simulate('blur')

    expect(wrapper.find('EditableNotes').prop('value')).toEqual('')
    expect(mockApi.setWfModuleNotes).toHaveBeenCalledWith(wfModule.id, '')

    expect(wrapper.find('.module-notes.visible')).toHaveLength(0)
  })

  it('queues changes from onChange and then submits them in onSubmit', () => {
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      parameter_vals: [
        { id: 1, parameter_spec: { id_name: 'a', type: 'string' }, value: 'A' },
        { id: 2, parameter_spec: { id_name: 'b', type: 'String' }, value: 'B' }
      ]
    }
    const wrapper = shallow(
      <WfModule
        {...props}
        wfModule={wfModule}
      />
    )

    wrapper.find('WfParameter[name="a"]').prop('onChange')('a', 'C')
    wrapper.update()
    expect(wrapper.find('WfParameter[name="a"]').prop('value')).toEqual('C')

    wrapper.find('WfParameter[name="b"]').prop('onChange')('b', 'D')
    wrapper.update()
    expect(wrapper.find('WfParameter[name="b"]').prop('value')).toEqual('D')

    // ... and neither should be submitted to the server
    expect(props.changeParam).not.toHaveBeenCalled()

    wrapper.find('WfParameter[name="b"]').prop('onSubmit')()
    expect(props.changeParam).toHaveBeenCalledWith(1, 'C')
    expect(props.changeParam).toHaveBeenCalledWith(2, 'D')

    // a bit of a white-box test: state should be cleared
    expect(wrapper.state('edits')).toEqual({})
  })

  it('resets just the one WfParmeter in onReset', () => {
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      parameter_vals: [
        { id: 1, parameter_spec: { id_name: 'a', type: 'string' }, value: 'A' },
        { id: 2, parameter_spec: { id_name: 'b', type: 'String' }, value: 'B' }
      ]
    }
    const wrapper = shallow(
      <WfModule
        {...props}
        wfModule={wfModule}
      />
    )

    wrapper.find('WfParameter[name="a"]').prop('onChange')('a', 'C')
    wrapper.find('WfParameter[name="b"]').prop('onChange')('b', 'D')
    wrapper.update()
    wrapper.find('WfParameter[name="b"]').prop('onReset')('b')
    wrapper.update()
    expect(wrapper.find('WfParameter[name="a"]').prop('value')).toEqual('C')
    expect(wrapper.find('WfParameter[name="b"]').prop('value')).toEqual('B')
  })

  it('submits a fake version_select click event in WfParameter onSubmit', () => {
    // Use case:
    // 1. User edits url field
    // 2. User clicks "submit" button within the URL field
    // Expected behavior: same as if user clicks "submit" button in the
    // WfParameter
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      parameter_vals: [
        { id: 1, parameter_spec: { id_name: 'url', type: 'string' }, value: '' },
        { id: 2, parameter_spec: { id_name: 'version_select', type: 'String' }, value: 'B' }
      ]
    }
    const wrapper = shallow(
      <WfModule
        {...props}
        wfModule={wfModule}
      />
    )

    wrapper.find('WfParameter[name="url"]').prop('onChange')('url', 'http://example.org')
    wrapper.find('WfParameter[name="url"]').prop('onSubmit')()

    expect(props.changeParam).toHaveBeenCalledWith(1, 'http://example.org')
    expect(mockApi.postParamEvent).toHaveBeenCalledWith(2)
  })

  it('submits a version_select click event in WfParameter[name=version_select] onSubmit', () => {
    // Use case: user wants to re-fetch
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      parameter_vals: [
        { id: 1, parameter_spec: { id_name: 'url', type: 'string' }, value: 'http://example.org' },
        { id: 2, parameter_spec: { id_name: 'version_select', type: 'String' }, value: 'B' }
      ]
    }
    const wrapper = shallow(
      <WfModule
        {...props}
        wfModule={wfModule}
      />
    )

    wrapper.find('WfParameter[name="version_select"]').prop('onSubmit')()

    expect(mockApi.postParamEvent).toHaveBeenCalledWith(2)
  })

  describe('lesson highlights', () => {
    let store
    let wrapper = null
    let nonce = 0

    function highlight (selectors) {
      lessonSelector.mockReturnValue({
        testHighlight: test => {
          return selectors.some(s => deepEqual(test, s))
        }
      })

      // trigger a change
      store.dispatch({ type: 'whatever', payload: ++nonce })
      if (wrapper !== null) wrapper.update()
    }

    beforeEach(() => {
      lessonSelector.mockReset()

      store = createStore((_, action) => ({
        workflow: { read_only: false, is_anonymous: false, wf_modules: [1, 2, 999] },
        wfModules: {
          999: { module_version: { module: 2 } }
        },
        modules: { 2: { name: 'Load from URL' } },
        ...action.payload
      }))

      highlight([])

      wrapper = mount(
        <Provider store={store}>
          <ConnectedWfModule {...props} />
        </Provider>
      )
    })
    afterEach(() => {
      wrapper.unmount()
      wrapper = null
    })

    it('highlights a WfModule', () => {
      highlight([{ type: 'WfModule', index: 2, moduleName: 'Load from URL' }])
      expect(wrapper.find('.wf-module').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights a WfModule', () => {
      // wrong name
      highlight([{ type: 'WfModule', index: 2, moduleName: 'TestModule2' }])
      expect(wrapper.find('.wf-module').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })

    it('highlights the "collapse" button', () => {
      highlight([{ type: 'WfModuleContextButton', index: 2, moduleName: 'Load from URL', button: 'collapse' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights the "collapse" button', () => {
      // wrong moduleName
      highlight([{ type: 'WfModuleContextButton', index: 2, moduleName: 'TestModule2', button: 'collapse' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).not.toMatch(/\blesson-highlight\b/)

      // wrong button
      highlight([{ type: 'WfModuleContextButton', index: 2, moduleName: 'Load from URL', button: 'notes' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })

    it('highlights the notes button', () => {
      highlight([{ type: 'WfModuleContextButton', index: 2, moduleName: 'Load from URL', button: 'notes' }])
      expect(wrapper.find('button.edit-note').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights the notes button', () => {
      // wrong moduleName
      highlight([{ type: 'WfModuleContextButton', index: 2, moduleName: 'TestModule2', button: 'notes' }])
      expect(wrapper.find('button.edit-note').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })
  })
})
