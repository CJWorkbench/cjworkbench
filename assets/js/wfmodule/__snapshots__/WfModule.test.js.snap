// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfModule, not read-only mode matches snapshot 1`] = `
<div
  className="wf-module expanded"
  data-module-name="Load from URL"
  onClick={[Function]}
>
  <div
    className="wf-card "
    draggable={true}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div>
      <div
        className="output-bar-container"
      >
        <StatusBar
          isDragging={false}
          isSelected={true}
        />
      </div>
      <div
        className="module-content"
      >
        <div
          className="module-card-header"
        >
          <WfModuleCollapseButton
            isCollapsed={false}
            isLessonHighlight={false}
            onCollapse={[Function]}
            onExpand={[Function]}
          />
          <i
            className="icon-undefined WFmodule-icon mr-2"
          />
          <div
            className="module-name"
          >
            Load from URL
          </div>
          <div
            className="context-buttons"
          >
            <a
              className="help-button"
              target="_blank"
              title="Help for this module"
            >
              <i
                className="icon-help"
              />
            </a>
            <button
              className="btn edit-note"
              onClick={[Function]}
              title="Edit Note"
            >
              <i
                className="icon-note"
              />
            </button>
            <WfModuleContextMenu
              id={999}
              removeModule={[Function]}
              stopProp={[Function]}
            />
          </div>
        </div>
        <StatusLine />
        <div
          className="module-card-params"
        >
          <WfParameter
            api={
              Object {
                "onParamChanged": [MockFunction],
                "postParamEvent": [MockFunction],
              }
            }
            changeParam={[Function]}
            getParamId={[Function]}
            getParamMenuItems={[Function]}
            getParamText={[Function]}
            isReadOnly={false}
            key="0"
            moduleName="Load from URL"
            p={
              Object {
                "id": 100,
                "parameter_spec": Object {
                  "id_name": "url",
                  "type": "string",
                },
                "value": "http://some.URL.me",
              }
            }
            revision={707}
            setClickNotification={[Function]}
            setParamText={[Function]}
            updateSettings={
              Object {
                "autoUpdateData": undefined,
                "lastUpdateCheck": undefined,
                "updateInterval": undefined,
                "updateUnits": undefined,
              }
            }
            wf_module_id={999}
          />
          <WfParameter
            api={
              Object {
                "onParamChanged": [MockFunction],
                "postParamEvent": [MockFunction],
              }
            }
            changeParam={[Function]}
            getParamId={[Function]}
            getParamMenuItems={[Function]}
            getParamText={[Function]}
            isReadOnly={false}
            key="1"
            moduleName="Load from URL"
            p={
              Object {
                "id": 101,
                "parameter_spec": Object {
                  "id_name": "version_select",
                  "type": "custom",
                },
              }
            }
            revision={707}
            setClickNotification={[Function]}
            setParamText={[Function]}
            updateSettings={
              Object {
                "autoUpdateData": undefined,
                "lastUpdateCheck": undefined,
                "updateInterval": undefined,
                "updateUnits": undefined,
              }
            }
            wf_module_id={999}
          />
        </div>
      </div>
      <div
        className="drop-alert "
      />
    </div>
  </div>
</div>
`;
